### PR TITLE
Environment setup

### PR DESCRIPTION
**Pull request by MarkHaoxiang**

### CODE REVIEW

**Comment by MarkHaoxiang** [comment]
Still very much a work-in-progress, but I think this is in a state that can be passed to main as a reference for frontend, as well as to separate off backend into separate branches (and start algorithms!)

File: backend/simulation/draw/visualiser.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        for edge in self.realm.edges.values():
+            self.draw_road(edge)
+
+
+    def draw_node(self, node: Node) -> None:
+        pygame.draw.circle(self.screen, "blue", node.pos.to_tuple(), 10)
+        for _ in node._trucks:
+            pygame.draw.circle(self.screen, "green", node.pos.to_tuple(), 2)
+
+    def draw_road(self, road: Edge) -> None:
```
Comment (sam1penny) [inline backend/simulation/draw/visualiser.py]: unsure why you changed the argument to an Edge and do an isinstance check?

File: backend/simulation/draw/visualiser.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from __future__ import annotations
+import pygame
+import sys
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from ..realm.graph import Node, Edge, Road
```
Comment (sam1penny) [inline backend/simulation/draw/visualiser.py]: with the `isinstance` check that is currently there, you need to import `Road` even if not type checking

File: backend/simulation/config.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import json
+
+class Config:
```
Comment (sam1penny) [inline backend/simulation/config.py:6]: We're only going to have one instance of Config ever, right? So perhaps a class is unnecessary here - a function to re-read the config might be more appropriate.

File: backend/simulation/config.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    
+    def __init__(self, path: str) -> None:
+        with open(path, 'r') as f:
+            self.data = json.load(f)
+
+    @property
+    def MAX_ACCELERATION(self):
```
Comment (sam1penny) [inline backend/simulation/config.py]: Type hints! (side note, I'll make a PR that makes mypy stricter)

File: shared/README.md [hunk 5]
```diff
@@ -1 +1 @@
-# Shared Files Will Live Here ()
\ No newline at end of file
+# Shared between both frontend and backend
```
Comment (sam1penny) [inline shared/README.md]: Did you mean to delete the specification for the intermediate representation?
It's important to keep it for the frontend and backend to communicate.

File: backend/simulation/env.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            rewards
+            dones
+            infos
+        """
+        self.config = Config(config_json_path)
+        self.realm = Realm(self.config)
+
+        return self.step()
+    
+    def step(self, actions: Dict[int, float] = {}):
```
Comment (sam1penny) [inline backend/simulation/env.py]: where do we compute actions?

**Review by sam1penny — COMMENTED** [review]
@MarkHaoxiang added some comments.

Btw, since we are frequently merging into environment, would you mind working on a separate branch next time?

File: backend/simulation/draw/visualiser.py (hunk 1 above)
Comment (MarkHaoxiang) [reply backend/simulation/draw/visualiser.py]: Leaves space to extend in the future.

File: backend/simulation/draw/visualiser.py (hunk 2 above)
Comment (MarkHaoxiang) [reply backend/simulation/draw/visualiser.py]: Will fix

File: backend/simulation/config.py (hunk 3 above)
Comment (MarkHaoxiang) [reply backend/simulation/config.py:6]: Was considering using config files to setup different scenarios for testing, as well as act as an intermediate data format if frontend wants to change a large part of the setup in the future. A config API here is more extendable, so perhaps we keep it in for now

File: backend/simulation/config.py (hunk 4 above)
Comment (MarkHaoxiang) [reply backend/simulation/config.py]: got it :D will fix!

File: shared/README.md (hunk 5 above)
Comment (MarkHaoxiang) [reply shared/README.md]: Leaving this as a TODO once intermediate representation is completely defined.

File: backend/simulation/env.py (hunk 6 above)
Comment (MarkHaoxiang) [reply backend/simulation/env.py]: Actions could be computed as an independent module.

ie. main.py
```
for step in range(128):
    actions = Algo.smart_algorithm(obs)
    obs, _, _, _ = env.step(actions)
```

File: shared/README.md (hunk 5 above)
Comment (sam1penny) [reply shared/README.md]: Hmm. I think this should be in the repo before we merge to main. afaik the intermediate representation is completely defined for us (and can be extended with front-end fields by their team)

File: backend/simulation/config.py (hunk 3 above)
Comment (sam1penny) [reply backend/simulation/config.py:6]: Fair enough. I still think we can achieve all of this without the need for creating a dummy object (can go into more detail if you want)

But anything is fine, up to you :)

File: backend/simulation/draw/visualiser.py (hunk 1 above)
Comment (sam1penny) [reply backend/simulation/draw/visualiser.py]: Ok - could you then rename the method and argument.

**Review by sam1penny — APPROVED** [review]
LGTM!

**Review by Phimaea — APPROVED** [review]

**Pull request merged**